FreeRTOS/FreeRTOS-Plus-TCP
# Bugfix for Socket lifetime issues

Bugfix for Socket lifetime issues

Have confirmed that this fix solves the issue originally reported in (#570)


Description
-----------

Added a new option to `vSocketClose` to select whether to destroy the socket at the end of close.  `bDestroy`
* Updated all uses of `vSocketClose`:
   * All user requested or new socket instances where it fails to create a new socket have `bDestroy = true`.
   * All internal close instances (such as in the TCP layer when far end is closed) have `bDestroy = false`.

Keep track of the lifetime of all sockets in 2 lists inside the FreeRTOS_Sockets.c file: `xActiveSocketsList` and `xInactiveSocketsList`.
Also check internally that the data arriving from events to the `IPTask` are valid and correct. - `socketASSERT_IS_VALID()`.
Check internally that only `vSocketClose` can be performed on a socket that hasn't been destroyed, and all other events/actions can only be performed on `active sockets`

Test Steps
-----------
See Related issue.
Confirmed that previous flow of multiple closes (by internal and user) are now handled properly: Log file of working solution:

```
Gain: Socket 12345 now has 1 / 5 child
prvSocketSetMSS: 1460 bytes for C0A80064ip:49222
Socket 12345 -> C0A80064ip:49222 State eCLOSED->eSYN_FIRST
prvWinScaleFactor: uxRxWinSize 512 MSS 1460 Factor 4
Socket 12345 -> C0A80064ip:49222 State eSYN_FIRST->eSYN_RECEIVED
TCP: passive 12345 => C0A80064ip:49222 set ESTAB (scaling 1)
Socket 12345 -> C0A80064ip:49222 State eSYN_RECEIVED->eESTABLISHED
Gain: Socket 12121 now has 1 / 5 child
prvSocketSetMSS: 1460 bytes for C0A80064ip:49223
Socket 12121 -> C0A80064ip:49223 State eCLOSED->eSYN_FIRST
prvWinScaleFactor: uxRxWinSize 512 MSS 1460 Factor 4
Socket 12121 -> C0A80064ip:49223 State eSYN_FIRST->eSYN_RECEIVED
TCP: passive 12121 => C0A80064ip:49223 set ESTAB (scaling 1)
Socket 12121 -> C0A80064ip:49223 State eSYN_RECEIVED->eESTABLISHED
Socket 12345 -> C0A80064ip:49222 State eESTABLISHED->eLAST_ACK
Socket 12121 -> C0A80064ip:49223 State eESTABLISHED->eLAST_ACK
Socket 12345 -> C0A80064ip:49222 State eLAST_ACK->eCLOSE_WAIT
vTCPStateChange: Closing socket
Socket 12121 -> C0A80064ip:49223 State eLAST_ACK->eCLOSE_WAIT
Lost: Socket 12121 now has 0 / 5 children
FreeRTOS_closesocket[12121 to C0A80064ip:49223]: buffers 64 socks 4
Lost: Socket 12345 now has 0 / 5 children
FreeRTOS_closesocket[12345 to C0A80064ip:49222]: buffers 64 socks 3
```

Related Issue
-----------
See Issue (#570)


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
